application/src/main/java/org/opentripplanner/transit/model/filter/transit/TripMatcherFactory.java

```diff
@@ -0,0 +1,64 @@
+package org.opentripplanner.transit.model.filter.transit;
+
+import java.time.LocalDate;
+import java.util.Set;
+import java.util.function.Function;
+import org.opentripplanner.transit.api.request.TripRequest;
+import org.opentripplanner.transit.model.filter.expr.ContainsMatcher;
+import org.opentripplanner.transit.model.filter.expr.EqualityMatcher;
+import org.opentripplanner.transit.model.filter.expr.ExpressionBuilder;
+import org.opentripplanner.transit.model.filter.expr.Matcher;
+import org.opentripplanner.transit.model.framework.FeedScopedId;
+import org.opentripplanner.transit.model.timetable.Trip;
+
+/**
+ * A factory for creating matchers for {@link Trip} objects.
+ * <p/>
+ * This factory is used to create matchers for {@link Trip} objects based on a request. The
+ * resulting matcher can be used to filter a list of {@link Trip} objects.
+ */
+public class TripMatcherFactory {
+
+  public static Matcher<Trip> of(
+    TripRequest request,
+    Function<FeedScopedId, Set<LocalDate>> serviceDateProvider
+  ) {
+    ExpressionBuilder<Trip> expr = ExpressionBuilder.of();
+
+    expr.atLeastOneMatch(request.agencies(), TripMatcherFactory::agencyId);
+    expr.atLeastOneMatch(request.routes(), TripMatcherFactory::routeId);
+    expr.atLeastOneMatch(
+      request.netexInternalPlanningCodes(),
+      TripMatcherFactory::netexInternalPlanningCode
+    );
+
+    expr.atLeastOneMatch(
+      request.serviceDates(),
+      TripMatcherFactory.serviceDate(serviceDateProvider)
+    );
+    return expr.build();
+  }
+
+  static Matcher<Trip> agencyId(FeedScopedId id) {
+    return new EqualityMatcher<>("agency", id, t -> t.getRoute().getAgency().getId());
+  }
+
+  static Matcher<Trip> routeId(FeedScopedId id) {
+    return new EqualityMatcher<>("route", id, t -> t.getRoute().getId());
+  }
+
+  static Matcher<Trip> netexInternalPlanningCode(String code) {
+    return new EqualityMatcher<>("privateCode", code, Trip::getNetexInternalPlanningCode);
+  }
+
+  static Function<LocalDate, Matcher<Trip>> serviceDate(
+    Function<FeedScopedId, Set<LocalDate>> serviceDateProvider
+  ) {
```

Comment:
You need to add doc on what the FeedScopedId id is in the function:

> Given a trip id the provider should return the service-date.

Reply:
Added comment.